### PR TITLE
Multi-server support & Generate ephemeral pubkeys randomly for each server

### LINKED ISSUE #9: Inter-server communication
The `derive_outputs` and `derive_kernels` apis (https://github.com/mimblewimble/mwixnet/blob/main/src/rpc.rs#L31) will be implemented along with the changes to [execute_round](https://github.com/mimblewimble/mwixnet/blob/main/src/server.rs#L171) necessary to consume them.

### LINKED ISSUE #10: e2e multi-server tests
Automated tests are needed that will spin up multiple mwixnet nodes (1...N) and perform swaps that test the entire multi-server flow.

### LINKED ISSUE #19: Generate random pubkey for each payload
The minimal space efficiency gains fail to justify the additional complexity of our current scheme. 
https://github.com/mimblewimble/mwixnet/pull/18#discussion_r1069533122

### PR DESCRIPTION
**Pull request by scilio**
Fixes https://github.com/mimblewimble/mwixnet/issues/9
Fixes https://github.com/mimblewimble/mwixnet/issues/10
Fixes https://github.com/mimblewimble/mwixnet/issues/19

### CODE REVIEW

File: doc/onion.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+}
+```
+
+The generated onion becomes:
+```json
+{
+    "commit": "0899dadc2b75d66d738b7dbfcba4a37460622dcedaf222e688a2a84826eaa1cff1",
+    "data": [
+		"d19f914e7a7b5ba1c0ab36d982bd0bc59aa651458a6ee86c2015b57d96424da4a41559069be5ba59e0d2b688f95f1dfb20648e21f9b01ac400cb4879f2ba434c8eb0337d3f58e9e643aa",
```
Comment (phyro) [inline doc/onion.md:151]: Is this entry a typo? If not, could you remind me what this is and how we pick or compute it?

File: doc/onion.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+For the first hop, the user provides:
+```json
+{
+  "server_pubkey": "96ced236bdf1aca722ef68b818445755e6ed4bacf23e19d7b71c43efc5f0077b",
+  "excess": "a9f15dc4760a1a280f68c6fc16d8aeada415fd66d5da805ff05cac6857a09db4",
+  "fee": 5,
+  "rangeproof": "None"
+}
+```
+The ephemeral key is randomly generated as
```
Comment (phyro) [inline doc/onion.md:128]: I think we're missing the second ephemeral key `5353ed848b8b2514aa08c8d9a5109ca4ddafe575c07a2a7cb2f19defa58d8442` here.

File: doc/store.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+## Data Model
+
+`SwapData`  entries are keyed with prefix 'S' followed by the commitment of the output being swapped. Entries are all unique by key.
+
+### `SwapData`
+
+The `SwapData` structure contains information needed to swap a single output. It has the following fields:
+
+- `excess`: The total excess for the output commitment.
```
Comment (phyro) [inline doc/store.md:17]: I'm assuming this isn't the total excess. If any of the mixnodes knows the total excess, they can link the inputs and outputs.

File: doc/store.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+- `output_commit`: The derived output commitment after applying excess and fee.
+- `rangeproof`: The rangeproof, included only for the final hop (node N).
+- `input`: The transaction input being spent.
+- `fee`: The transaction fee.
+- `onion`: The remaining onion after peeling off our layer.
```
Comment (phyro) [inline doc/store.md:22]: Out of curiosity, would it not be simpler to save only the starting onion and the status? Then the first node just  needs to peel the layer like any other.

File: src/tx.rs [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+	wallet: &Arc<dyn Wallet>,
+	components: &TxComponents,
+	output_excesses: &Vec<SecretKey>,
+	fee_base: u64,
+	fees_paid: u64,
+) -> Result<TxComponents, TxError> {
+	// calculate minimum fee required for the kernel
+	let min_kernel_fee = TransactionBody::weight_by_iok(0, 0, 1) * fee_base;
+
+	add_kernel_and_collect_fees(
```
Comment (phyro) [inline src/tx.rs:92]: I remember you mentioned you wanted to make it a single kernel as described [here](https://forum.grin.mw/t/mimblewimble-coinswap-proposal/8322/57). Is there a reason we don't use this optimization?

**Comment by phyro** [comment]
I think we should expose relevant things like crypto, onions, API structs and possibly more as a lib to make it possible for other software (wallet?) to easily use the onion and other functionality.

File: src/servers/swap.rs [hunk 6]
```diff
@@ -1,687 +1,807 @@
...
+	fn swap(&self, onion: &Onion, comsig: &ComSignature) -> Result<(), SwapError>;
+
+	/// Iterate through all saved submissions, filter out any inputs that are no longer spendable,
+	/// and assemble the coinswap transaction, posting the transaction to the configured node.
+	fn execute_round(&self) -> Result<Option<Transaction>, Box<dyn std::error::Error>>;
+}
+
+/// The standard MWixnet server implementation
+#[derive(Clone)]
+pub struct SwapServerImpl {
```
Comment (phyro) [inline src/servers/swap.rs:60]: I'm still battling through this so not all things are clear in my head. I find it a bit complex that we have two different server implementations when the only difference is that one has a facade `swap` api exposed and stores some things in memory. Was there a specific reason you decided to separate the two and not have `Option<SwapImpl>` or similar on the mix server?

**Comment by phyro** [comment]
@scilio given our limited capabilities to review, I think a good next step would be to make it possible to test this on a testnet. As said before, this would require us to have the ability to use the code to generate hops, onions, comsig on grin-wallet. It's also a good way to test if mixnodes creating their fee outputs works.

**Comment by scilio** [comment]
> @scilio given our limited capabilities to review, I think a good next step would be to make it possible to test this on a testnet. As said before, this would require us to have the ability to use the code to generate hops, onions, comsig on grin-wallet. It's also a good way to test if mixnodes creating their fee outputs works.

Integration with grin-wallet is potentially a non-trivial task. I should be able to refactor the onion generation code into its own library now, but I haven't fully wrapped my head around the grin-wallet design enough yet to know how complicated integration would be. I was hoping to explore that idea more and possibly submit a funding request for a milestone 4 to do that integration.

**Comment by yeastplume** [comment]
I'm merging this now to make the code easier to work with for a basic wallet integration. PR is still available for review.

**Pull request merged**